components/org.wso2.identity.event.websubhub.publisher/src/main/java/org/wso2/identity/event/websubhub/publisher/service/WebSubEventSubscriberImpl.java

```diff
@@ -0,0 +1,195 @@
+            return false;
+        }
+    }
+
+    private void makeSubscriptionAPICall(String topic, String webSubHubBaseUrl, String operation, String callbackUrl)
+            throws WebSubAdapterException {
+
+        String subscriptionUrl = buildSubscriptionURL(topic, webSubHubBaseUrl, operation, callbackUrl);
+
+        ClientManager clientManager = WebSubHubAdapterDataHolder.getInstance().getClientManager();
+        HttpPost httpPost = clientManager.createHttpPost(subscriptionUrl, null);
+
+        WebSubHubCorrelationLogUtils.triggerCorrelationLogForRequest(httpPost);
+        final long requestStartTime = System.currentTimeMillis();
+
+        CompletableFuture<HttpResponse> future = clientManager.executeAsync(httpPost);
+
+        future.thenAccept(response -> {
+            try {
+                handleSubscriptionResponse((CloseableHttpResponse) response, httpPost, topic, operation,
+                        requestStartTime);
+            } catch (IOException | WebSubAdapterException e) {
+                log.error("Handling subscription response failed. ", e);
+            }
+        }).exceptionally(ex -> {
+            log.error("Subscription API call failed. ", ex);
+            return null;
+        });
+    }
+
+    private void handleSubscriptionResponse(CloseableHttpResponse response, HttpPost httpPost,
+                                            String topic, String operation, long requestStartTime)
+            throws IOException, WebSubAdapterException {
+
+        StatusLine statusLine = response.getStatusLine();
+        int responseCode = statusLine.getStatusCode();
+        String responsePhrase = statusLine.getReasonPhrase();
+
+        if (responseCode == HttpStatus.SC_OK) {
+            HttpEntity entity = response.getEntity();
+            WebSubHubCorrelationLogUtils.triggerCorrelationLogForResponse(httpPost, requestStartTime,
+                    WebSubHubCorrelationLogUtils.RequestStatus.COMPLETED.getStatus(),
+                    String.valueOf(responseCode), responsePhrase);
+            handleSuccessfulOperation(entity, topic, operation);
+        } else if (responseCode == HttpStatus.SC_NOT_FOUND) {
+            HttpEntity entity = response.getEntity();
+            WebSubHubCorrelationLogUtils.triggerCorrelationLogForResponse(httpPost, requestStartTime,
+                    WebSubHubCorrelationLogUtils.RequestStatus.FAILED.getStatus(),
+                    String.valueOf(responseCode), responsePhrase);
+            handleErrorResponse(entity, topic, operation);
+        } else {
+            WebSubHubCorrelationLogUtils.triggerCorrelationLogForResponse(httpPost, requestStartTime,
+                    WebSubHubCorrelationLogUtils.RequestStatus.CANCELLED.getStatus(),
+                    String.valueOf(responseCode), responsePhrase);
+            HttpEntity entity = response.getEntity();
+            handleFailedOperation(entity, topic, operation, responseCode);
```

Comment:
For each of these we need to have diagnostic logs

Reply:
Should we handle these response separately ? I think we can move this to a proper client implementation and re use. In such cases we can even add retry mechanism as well

Reply:
Create a separate issue for this refactoring.

Reply:
https://github.com/wso2-extensions/identity-event-publishers/pull/13#discussion_r2092646106